File: Week 05/id_576/Fibonacci.java
/**
 * Fibonacci
 */
public class Fibonacci {

    int fib1 (int n) {
        if (n <= 0) {
            return 0;
        } else if (n == 1) {
            return 1;
        } else{
            return fib(n - 1) + fib(n - 2);
        }
    }

    //更简洁的写法
    int fib2 (int n) {
        return n <= 1 ? n : fib1(n - 1) + fib1(n - 2);
    }

    //时间复杂度上的优化，添加一个缓存,o(n)， 记忆化搜索
    int fib3 (int n, int[] memo) {
        //terminator
        if (n <= 1) {
            return n;
        } 
        if (memo[n] == 0) {
            memo[n] = fib2(n - 1) + fib2(n - 2);
        }
        return memo[n];
    }

    //bottom up，终极模版
    int fib4 (int n) {
        if (n <= 1) return n;
        int[] a = new int[n];
        a[0] = 0;
        a[1] = 1;
        for (int i = 2; i <= n; i++) {
            a[i] = a[i - 1] + a[i - 2];
        }
        return a[n];
    }
}

File: Week 05/id_441/fibonacci.js
// 斐波那契代码
function fib (n, memo) {
    if( n < 0) {
        return 0
    }else if(n === 1) {
        return 1
    }else if(memo[n] === 0){
        memo[n] = fib(n-1) + fib(n-2)
    }
    return memo[n]
}
function fib (n, memo) {
    if(n<=1) {
        return n;
    }
    if(!memo[n]){
        memo[n] = fib(n-1)+fib(n-2)
    }
    return memo[n]
}
function fib (n){
    let a = [0,1]
    console.log()
    for(let i =2; i<n; ++i){
        console.log(a[i])
        a[i] = a[i-1]+a[i-2]
    }
    return a[n]
}
fib(3)

// 二维的fib格子走法
// 一维的递推： f(n) = f(n-1) + f(n-2)
// 二维的递推公式： f(i,j) = f(i-1,j)+f(i,j-1)
function fib(m,n){
    // 初始化x轴和y轴
    let a = [[],[]]
    for(let i =0 ; i<10; ++i) {
        a.push([])
        a[i][0] = 1
    }
    for(let j =0; j<10; ++j) {
        a[0][j] = 1
    }
    console.log(a)
    for(let  j=0; j<m;  ++j){
        for(let k=0; k<n; ++k){
            a[j][k] = a[j-1][k] + a[j][k-1]
        }
    }
    return a[m-1][n-1]
}

let a = [[],[]]
    for(let i =0 ; i<10; ++i) {
        a.push([])
        a[i][0] = 1
    }
    for(let j =0; j<10; ++j) {
        a[0][j] = 1
    }
    console.log(a)
